### PR TITLE
Add implementation of `adjustsFontSizeToFit` on the new architecture on Android

### PR DESCRIPTION
**Pull request by j-piasecki**
## Summary:

`adjustsFontSizeToFit` prop is exposed on both platforms but is missing implementation on the new arch on Android. This Pr adds it.

Fixes https://github.com/facebook/react-native/issues/43104

## Changelog:

[ANDROID] [FIXED] - Fixed `adjustsFontSizeToFit`  not working on Android when using the new architecture

## Test Plan:

Tested on the RN Tester app using the `AdjustingFontSize` test:

| Old architecture | New architecture |
|------------------|------------------|
| <video src="https://github.com/facebook/react-native/assets/21055725/9243317c-1917-4ddb-9b8a-e9e99638409d">           | <video src="https://github.com/facebook/react-native/assets/21055725/39a7a9f2-21e4-4ba7-9ceb-dfec4ca6f643">         |

### CODE REVIEW

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java [hunk 1]
```diff
@@ -407,6 +459,37 @@ public static long measureText(
...
+        ? paragraphAttributes.getInt(PA_KEY_MAX_NUMBER_OF_LINES)
+        : ReactConstants.UNSET;
+
+    if (adjustFontSizeToFit) {
+      double minimumFontSize = paragraphAttributes.contains(PA_KEY_MINIMUM_FONT_SIZE)
+        ? paragraphAttributes.getDouble(PA_KEY_MINIMUM_FONT_SIZE)
+        : Double.NaN;
+      double maximumFontSize = paragraphAttributes.contains(PA_KEY_MINIMUM_FONT_SIZE)
+        ? paragraphAttributes.getDouble(PA_KEY_MINIMUM_FONT_SIZE)
+        : Double.NaN;
```
Comment (cortinico) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java:477]:
```suggestion
      double maximumFontSize = paragraphAttributes.contains(PA_KEY_MAXIMUM_FONT_SIZE)
        ? paragraphAttributes.getDouble(PA_KEY_MAXIMUM_FONT_SIZE)
        : Double.NaN;
```

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java [hunk 2]
```diff
@@ -378,6 +382,54 @@ private static Layout createLayout(
...
   }
 
+  public static void adjustSpannableFontToFit(
+    Spannable text,
+    float width,
+    YogaMeasureMode widthYogaMeasureMode,
+    float height,
+    YogaMeasureMode heightYogaMeasureMode,
+    double minimumFontSizeAttr,
+    double maximumFontSizeAttr,
```
Comment (cortinico) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java:392]: This is unused

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java [hunk 3]
```diff
@@ -378,6 +382,54 @@ private static Layout createLayout(
...
+    // Find the largest font size used in the spannable to use as a starting point.
+    // Minimum font size is 4pts to match the iOS implementation.
+    int currentFontSize = (int) (Double.isNaN(minimumFontSizeAttr) ? PixelUtil.toPixelFromDIP(4) : minimumFontSizeAttr);
+    ReactAbsoluteSizeSpan[] spans = text.getSpans(0, text.length(), ReactAbsoluteSizeSpan.class);
+    for (ReactAbsoluteSizeSpan span : spans) {
+      currentFontSize = Math.max(currentFontSize, span.getSize());
+    }
+
+    int initialFontSize = currentFontSize;
+    int minimumFontSize = (int) (Double.isNaN(minimumFontSizeAttr) ? PixelUtil.toPixelFromDIP(4) : minimumFontSizeAttr);
```
Comment (cortinico) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java:409]: This is practically duplicated from line 402. Could we avoid doing this check twice?

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java [hunk 4]
```diff
@@ -378,6 +382,54 @@ private static Layout createLayout(
...
+    while (currentFontSize > minimumFontSize
+      && (maximumNumberOfLines != ReactConstants.UNSET && layout.getLineCount() > maximumNumberOfLines
+      || heightYogaMeasureMode != YogaMeasureMode.UNDEFINED && layout.getHeight() > height)) {
```
Comment (cortinico) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java:412]:
```suggestion
        while (currentFontSize > minimumFontSize
        && ((maximumNumberOfLines != ReactConstants.UNSET
                && layout.getLineCount() > maximumNumberOfLines)
            || (heightYogaMeasureMode != YogaMeasureMode.UNDEFINED
                && layout.getHeight() > height))) {
```

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java [hunk 5]
```diff
@@ -590,6 +668,8 @@ public static WritableArray measureLines(
         TextAttributeProps.getTextBreakStrategy(
             paragraphAttributes.getString(PA_KEY_HYPHENATION_FREQUENCY));
 
+    // TODO: adjustFontSizeToFit
```
Comment (cortinico) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java:671]: This is missing?

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java (hunk 3 above)
Comment (j-piasecki) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java]: Done in https://github.com/facebook/react-native/pull/44075/commits/4451253f2cde47afe666bf60ccb97642960faf33

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java (hunk 2 above)
Comment (j-piasecki) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java]: Removed in https://github.com/facebook/react-native/pull/44075/commits/367f65bad5a8521c02e1d63ab582d1d3c1668ecd

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java (hunk 5 above)
Comment (j-piasecki) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java]: Yes, I didn't want to repeat everything twice while testing, added in https://github.com/facebook/react-native/pull/44075/commits/e2bba84a5c63564fbb439722cbac8f49995b05e0.

**Comment by cortinico** [comment]
I've done some testing of this feature on RN-Tester and on Fabric the text is cropped:

| Old Arch | New Arch (this PR) |
| -------- | ------------------ |
| before | after |

**Comment by j-piasecki** [comment]
@cortinico Which commit were you testing on? This looks like before https://github.com/facebook/react-native/pull/44075/commits/7c2f944662b51be0fd47474a7a44236c8b10856c.

The behavior now looks ok to me:

https://github.com/facebook/react-native/assets/21055725/f9574f15-86b7-41d1-8cdb-c8910766c036

**Comment by cortinico** [comment]
> 7579c96

I was on [7579c96](https://github.com/facebook/react-native/pull/44075/commits/7579c96059e294863cbfb748d2c908c628158daa) yeah. I'll have to re-import and re-test this 👍 Let me get back to you

**Comment by cortinico** [comment]
I've just re-tested it and it works great now 👍 @j-piasecki

**Review by cortinico — APPROVED** [review]

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java [hunk 6]
```diff
@@ -648,6 +681,7 @@ public void setBorderStyle(@Nullable String style) {
 
   public void setSpanned(Spannable spanned) {
     mSpanned = spanned;
+    mShouldAdjustSpannableFontSize = true;
```
Comment (cortinico) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java:684]: QQ: Should we set this to true also inside:

- `setNumberOfLines`
- `setMinimumFontSize`

and theoretically also inside the following overrides (which we'll have to add):

- `setIncludeFontPadding()`
- `setBreakStrategy()`
- `setHyphenationFrequency()`

As they all contribute to the computation of `adjustSpannableFontToFit`

**Review by cortinico — APPROVED** [review]
This is good to go @j-piasecki 🎉  Just one question left

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java (hunk 6 above)
Comment (j-piasecki) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java:704]: I believe you're right, done in 1c879ac

**Pull request closed without merging**